1.segment_images/scripts/4.reconstruct_3D_masks.py

```diff
@@ -137,256 +171,165 @@
-print(results_df.shape)
-results_df.head()
-
-
-# In[12]:
-
-
-# add an object number to each row
-results_df.reset_index(drop=True, inplace=True)
-results_df["object_number"] = results_df.index
-results_df.drop(columns=["node1", "node2"], inplace=True)
-# un group the path column
-results_df = results_df.explode("path")
-results_df.rename(columns={"path": "node"}, inplace=True)
-
-
-# In[13]:
-
-
-# # get the x and y coordinates of each node from the original df
-results_df["coorinate-0"] = results_df["node"].apply(
-    lambda x: coordinates_df[coordinates_df["unique_id"] == x]["centroid-0"].values[0]
-)
-results_df["coorinate-1"] = results_df["node"].apply(
-    lambda x: coordinates_df[coordinates_df["unique_id"] == x]["centroid-1"].values[0]
-)
-results_df["slice"] = results_df["node"].apply(
-    lambda x: coordinates_df[coordinates_df["unique_id"] == x]["slice"].values[0]
-)
-results_df["label"] = results_df["node"].apply(
-    lambda x: coordinates_df[coordinates_df["unique_id"] == x]["label"].values[0]
-)
-print(results_df.shape)
-results_df
-
-
-# ## Plot paths output
-
-# In[14]:
-
-
-if in_notebook:
-    # plot the data in 3D with tracks
-    fig = plt.figure()
-    ax = fig.add_subplot(111, projection="3d")
-    ax.scatter(
-        results_df["coorinate-0"],
-        results_df["coorinate-1"],
-        results_df["slice"],
-        c=results_df["object_number"],
-    )
-    ax.set_xlabel("X Label")
-    ax.set_ylabel("Y Label")
-    ax.set_zlabel("Z Label")
-    plt.show()
-
-
-# ## Generate the new image via mask number reassignment
-
-# In[16]:
-
-
-# go back through the image and color each mask with the object number based on x and y coordinates for each slice
-# create a new image with the same shape as the original image
-new_image = np.zeros_like(image)
-# iterate over each slice
+new_mask_image = np.zeros_like(image)
+# mask label reassignment
 for slice in range(image.shape[0]):
-    # relate the object number to each mask
-    for index, row in results_df[results_df["slice"] == slice].iterrows():
-        # replace the mask label with the object number
-        new_image[slice][image[slice] == row["label"]] = row["object_number"]
-
-
-# In[17]:
-
-
-# rescale the images so that the max pixel value is 255
-new_image = (new_image / np.max(new_image)) * 255
-# conver image to uint8 int
-new_image = new_image.astype(np.uint8)
-
-
-# In[18]:
-
+    mask = image[slice, :, :]
+    tmp_df = return_df[return_df["slice"] == slice]
+    for i in range(tmp_df.shape[0]):
+        mask[mask == tmp_df.iloc[i]["original_label"]] = tmp_df.iloc[i]["new_label"]
 
+    new_mask_image[slice, :, :] = mask
 # save the new image
-tifffile.imwrite(output_image_dir, new_image)
+tifffile.imwrite(output_image_dir, new_mask_image)
 
 
-# ## Visualize the new image per z-slice
-
-# In[19]:
+# In[ ]:
 
 
-# plot the new image masks
 if in_notebook:
-    for slice in range(new_image.shape[0]):
+    for slice in range(new_mask_image.shape[0]):
```

Comment:
What is the difference between the original and new mask image? From the montage in the notebook, both look the same.